### PR TITLE
feat: improve syntax error messages

### PR DESCRIPTION
**Pull request by aliehsaeedii**
### Description 
Antlr error messages are customised in order to address syntax errors in a more informative way.
### Testing done 
Unit tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java [hunk 1]
```diff
@@ -127,16 +134,65 @@ public void syntaxError(
...
+
+  private String getExpectations(
+      final String message,
+      final String offendingToken,
+      final int line,
+      final int charPositionInLine) {
+    final StringBuilder output = new StringBuilder();
+    final String expectingStr = message.split("expecting ")[1];
+    // If the command is mistyped, fins the most similar command and show it
+    if (line == 1 && charPositionInLine == 0) { // this is a command typo
```
Comment (jzaralim) [inline ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java:169]: The position of the first word in a command is not necessarily at (line 1:0). For example, if you try running `show streams; shw streams;` the error message you get is `line 1:13: extraneous input 'shw' expecting ...`

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java [hunk 2]
```diff
@@ -43,6 +47,9 @@ public class SyntaxErrorValidator extends BaseErrorListener {
       "extraneous input.*expecting.*");
   private static final Pattern MISMATCHED_INPUT_PATTERN = Pattern.compile(
       "mismatched input.*expecting.*");
+  private final List<String> commands = Arrays.asList("SELECT", "CREATE", "INSERT", "DESCRIBE",
```
Comment (jzaralim) [inline ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java:50]: Would it be possible to use the `GrammarTokenExporter` instead? It returns more tokens than just the commands, but it would prevent people from having to update extra lists when they add new syntax to ksqlDB.

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java [hunk 3]
```diff
@@ -127,16 +134,65 @@ public void syntaxError(
...
+    final String expectingStr = message.split("expecting ")[1];
+    // If the command is mistyped, fins the most similar command and show it
+    if (line == 1 && charPositionInLine == 0) { // this is a command typo
+      output.append(String.format("Unknown statement '%s'\n",
+              offendingToken));
+      output.append(
+              String.format("Did you mean '%s'?", getMostSimilar(offendingToken)));
+    // In case of missing closing brackets or parentheses
+    } else if (message.contains("EOF")) {
+      output.append("Syntax error at or near \";\"\n");
```
Comment (jzaralim) [inline ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java:176]: Would a statement that's missing a semicolon end up here? If it does, then the error message won't make sense.

Comment (aliehsaeedii) [reply ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java:176]: > Would a statement that's missing a semicolon end up here? If it does, then the error message won't make sense.

No. It does not.

**Comment by spena** [comment]
@aliehsaeedii I executed a couple of statements and got these errors:

One
```
ksql> crate stream s1(id int) with (kafka_topic='s1');
line 1:1: Syntax Error
Unknown statement 'crate'
Did you mean 'PRINT'?
Statement: crate stream s1(id int) with (kafka_topic='s1');
Caused by: line 1:1: Syntax Error
Unknown statement 'crate'
Did you mean
	'PRINT'?
Caused by: org.antlr.v4.runtime.InputMismatchException
```

Two
```
ksql> show streams; sho streams;
line 1:15: Syntax Error
Unknown statement 'sho'
Did you mean 'SET'?
Statement: show streams; sho streams;
Caused by: line 1:15: Syntax Error
Unknown statement 'sho'
Did you mean 'SET'?
```

I don't know why the error is displayed twice. I don't see that in your code. I think we should figure out how to avoid the duplicate message so we have a cleaner error message. 

Now, the errors below didn't guess what the user meant. Instead of `SHOW`, it uses `SET`. Let's skip the use of the Levenshtein Distance algorithm for now. There are some tweaks we need to do to the algorithm in order to get more accurate results. 
```
ksql> crate stream s1(id int) with (kafka_topic='s1');
line 1:1: Syntax Error
Unknown statement 'crate'
Did you mean 'PRINT'?

ksql> show streams; sho streams;
line 1:15: Syntax Error
Unknown statement 'sho'
Did you mean 'SET'?
```

There are some errors that are not improved. i.e.
```
ksql> show stream;
line 1:6: Syntax Error
no viable alternative at input 'show stream'
Statement: show stream;

ksql> create tream s1(id int) with (kafka_topic='s1');
line 1:8: Syntax Error
no viable alternative at input 'create tream'
Statement: create tream s1(id int) with (kafka_topic='s1');
```

What about displaying an error like this?
```
ksql> create tream s1(id int) with (kafka_topic='s1');
ERROR: You have a syntax error at or near 'tream' on line 1:7.
Line 1: create tream s1(id int) with (kafka_topic='s1');
               ^
```

Other examples:
```
ksql> show streams; sho streams;
ERROR: You have a syntax error at or near 'sho' on line 1:14.
Line 1: show streams; sho streams;
                      ^
```

```
ksql> show
>stream;
ERROR: You have a syntax error at or near 'stream' on line 2:1.
Line 2: stream;
        ^
```

**Comment by aliehsaeedii** [comment]
> test: update the error message in unit test

**Comment by aliehsaeedii** [comment]
> I don't know why the error is displayed twice. I don't see that in your code. I think we should figure out how to avoid the duplicate message so we have a cleaner error message.

I think it is better to solve this issue [here](https://github.com/confluentinc/ksql/pull/9531).

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java [hunk 4]
```diff
@@ -127,16 +135,77 @@ public void syntaxError(
...
+      output.append(String.format("Unknown statement '%s'%n", offendingToken));
+      output.append(
+              String.format("Did you mean '%s'?", getMostSimilar(offendingToken)));
+    // In case of missing closing brackets or parentheses
+    } else if (message.contains("EOF")) {
+      output.append("Syntax error at or near \";\"\n");
+      output.append("'}', ']', or ')' is missing");
+    // If the expecting list generated by Antlr is a short one, show it to user
+    } else if (expectingStr.split(",").length <= 3) {
+      output.append(String.format("Expecting %s", expectingStr));
```
Comment (jzaralim) [inline ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java:184]: This could be in another PR, but the expecting list generated by Antlr is often pretty bad. For example,

```
ksql> describe functiii concat;
line 1:19: Syntax Error
Expecting ';'
Statement: describe functiii concat;
Caused by: line 1:19: Syntax Error
Expecting ';'

ksql> create stream foo a select * from bar;
line 1:19: Syntax Error
Expecting {';', '(', 'WITH'}
Statement: create stream foo a select * from bar;
Caused by: line 1:19: Syntax Error
Expecting {';', '(', 'WITH'}
Caused by: org.antlr.v4.runtime.InputMismatchException

```

**Review by jzaralim — APPROVED** [review]
LGTM. One comment, but it's not a regression so I'd be okay with merging this.
